File: 3546. equal-sum-grid-partition-i.java
import java.util.*;

class Solution {
    int n, m;

    boolean checkHz(int[][] grid){
        HashSet<Long> st = new HashSet<>();

        long pref = 0;

        for(int i = 0; i < n; i++){
            long rowSum = 0;
            for(int j = 0; j < m; j++){
                rowSum += grid[i][j];
            }
            pref += rowSum;
            st.add(pref);
        }

        if(pref % 2 != 0) return false;

        return st.contains(pref / 2);
    }

    boolean checkVz(int[][] grid){
        HashSet<Long> st = new HashSet<>();

        long pref = 0;

        for(int j = 0; j < m; j++){
            long colSum = 0;
            for(int i = 0; i < n; i++){
                colSum += grid[i][j];
            }
            pref += colSum;
            st.add(pref);
        }

        if(pref % 2 != 0) return false;

        return st.contains(pref / 2);
    }

    public boolean canPartitionGrid(int[][] grid) {
        n = grid.length;
        m = grid[0].length;

        return checkHz(grid) || checkVz(grid);
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        int[][] grid1 = {{1,4},{2,3}};
        boolean output1 = sol.canPartitionGrid(grid1);
        System.out.println("Example 1: grid=[[1,4],[2,3]]");
        System.out.println("Output: " + output1);
        System.out.println("Expected: true");
        System.out.println();
        
        // Example 2
        int[][] grid2 = {{1,3},{2,4}};
        boolean output2 = sol.canPartitionGrid(grid2);
        System.out.println("Example 2: grid=[[1,3],[2,4]]");
        System.out.println("Output: " + output2);
        System.out.println("Expected: false");
    }
}

File: 3546. equal-sum-grid-partition-i.py
class Solution(object):
    def canPartitionGrid(self, grid):
        n = len(grid)
        m = len(grid[0])

        def checkHz():
            st = set()
            pref = 0

            for i in range(n):
                rowSum = 0
                for j in range(m):
                    rowSum += grid[i][j]
                pref += rowSum
                st.add(pref)

            if pref % 2 != 0:
                return False

            return (pref // 2) in st

        def checkVz():
            st = set()
            pref = 0

            for j in range(m):
                colSum = 0
                for i in range(n):
                    colSum += grid[i][j]
                pref += colSum
                st.add(pref)

            if pref % 2 != 0:
                return False

            return (pref // 2) in st

        return checkHz() or checkVz()

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    grid1 = [[1,4],[2,3]]
    output1 = sol.canPartitionGrid(grid1)
    print(f"Example 1: grid=[[1,4],[2,3]]")
    print(f"Output: {output1}")
    print(f"Expected: True")
    print()
    
    # Example 2
    grid2 = [[1,3],[2,4]]
    output2 = sol.canPartitionGrid(grid2)
    print(f"Example 2: grid=[[1,3],[2,4]]")
    print(f"Output: {output2}")
    print(f"Expected: False")